app/controllers/time_slot_controller.rb

```diff
@@ -52,40 +51,24 @@ def cancel
   end
 
   def index
-    @appointment = current_patient.current_appointment
-    @ubs = @appointment.try(:ubs)
-
-    @patient_can_schedule = current_patient.can_schedule?
-
     return render_vaccinated if current_patient.vaccinated?
 
-    @gap_in_days = slot_params[:gap_in_days].to_i || 0
-    @current_day = Time.zone.now + @gap_in_days.days
-
-    @time_slots = {}
+    @appointment = current_patient.current_appointment
+    @gap_in_days = gap_in_days
+    @current_day = [Time.zone.now.at_beginning_of_day + @gap_in_days.days, Time.zone.now].max # prevent users from scheduling in the past
+    @patient_can_schedule = current_patient.can_schedule?
 
     last_appointment = current_patient.last_appointment
     return if last_appointment&.second_dose? && @current_day.to_date < last_appointment.start.to_date
 
-    Ubs.where(active: true).each do |ubs|
-      slots = []
-
-      if @gap_in_days <= TimeSlotController::SLOTS_WINDOW_IN_DAYS && @gap_in_days >= 0
-        if @current_day.today?
-          appointments = Appointment.where(start: @current_day..@current_day.end_of_day, ubs: ubs, patient_id: nil)
-        else
-          appointments = Appointment.where(start: @current_day.at_beginning_of_day..@current_day.end_of_day, ubs: ubs, patient_id: nil)
-        end
-
-        next unless appointments.exists?
-
-        appointments.each do |appointment|
-          slots << { slot_start: appointment.start, slot_end: appointment.end }
-        end
-
-        @time_slots[ubs] = slots.uniq
-      end
-    end
+    @time_slots = Appointment.
+      includes(:ubs).
+      free. # can be scheduled
+      start_between(@current_day, @current_day.end_of_day). # in the date the user is looking for
+      order(:start). # in chronological order
+      select(:ubs_id, :start, :end). # only return what we care with
+      distinct. # remove duplicates (same as .uniq in pure Ruby)
+      group_by { |a| a.ubs } # transforms it into a Hash grouped by Ubs
```

Comment:
Main start of this PR: uses a single SQL query to fetch all free Appointments along with Ubs.